### PR TITLE
SignUp

### PR DESCRIPTION
**Pull request by TuanNhon**

### CODE REVIEW

**Comment by TuanNhon** [comment]
ready

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def index
+    @users = User.all
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: - select các trường cần hiển thị
- sắp xếp bản ghi
- phân trang

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    @user = User.new user_params
+    if @user.save
+      redirect_to @user
+    else
+      render :new
+    end
+  end
+
+  def show
+    @user = User.find_by id: params[:id]
```
Comment (huongnguyenmta) [inline app/controllers/users_controller.rb]: check `nil`

**Comment by huongnguyenmta** [comment]
commented
Có mí cái lỗi mà cứ để comment hoài à (lengoi)(songphi)

**Comment by TuanNhon** [comment]
ready

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    else
+      render :new
+    end
+  end
+
+  def show
+    unless @user = User.find_by(id: params[:id])
+      flash[:danger] = t ".notfound"
+      redirect_to users_path
+    end
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]:
```ruby
@user = User.find_by id: params[:id] 
return if @user 
flash[:danger] = t ".not_found"
redirect_to users_path
```

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  scope :selected, -> {select :id, :name, :email}
+  scope :ordered, -> {order name: :asc}
+  validates :name,  presence: true,
+    length: {maximum: Settings.user_valid.name_length}
+  validates :email, presence: true,
+    length: {maximum: Settings.user_valid.email_length},
+    format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
+  validates :password, presence: true,
+    length: {minimum: Settings.user_valid.password_length}
+  has_secure_password
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Add the codes in following order
```ruby
 #all mixins
   include Something
   extend Something

   #other stuff
   acts_as_taggable
   paginates

   #associations
   has_many :something
   belongs_to :something_else

   #validations
   validate_presence_of :something

   #scopes
   scope :something

   #instance methods
   def instance_method
   end

   #class methods
   def self.method
   end

   #private methods
   private
   def method2
   end
```

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class User < ApplicationRecord
+  before_save {email.downcase!}
+  VALID_EMAIL_REGEX = Settings.user_valid.email_syntax
+  scope :selected, -> {select :id, :name, :email}
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove unnecessary white space
`scope :selected, ->{select :id, :name, :email}`

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  scope :ordered, -> {order name: :asc}
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: The same here

File: app/views/users/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+<h2><%= t ".list" %></h2>
+<table>
+  <% @users.each do |user| %>
+    <tr>
+      <td>
+        <%= user.id %>
+      </td>
+      <td>
+        <%= user.name %> --
```
Comment (mdmahmudurrahman) [inline app/views/users/index.html.erb]: What is "--" here? :)

File: config/locales/en.yml [hunk 8]
```diff
@@ -1,33 +1,12 @@
...
+    new:
+      submit: "Create my account"
+      signup: "SignUp"
+    index:
+      list: "List users:"
+    form:
+      name: "Name"
+      email: "Email"
+      password: "Password"
+    notfound: "User not found"
```
Comment (mdmahmudurrahman) [inline config/locales/en.yml]: Use a meaningful and understandable key name
`not_found`

File: config/routes.rb [hunk 9]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get "/signup", to: "users#new"
+  post "/signup", to: "users#create"
+  resources :users
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: Specify the action here for this resource

**Comment by mdmahmudurrahman** [comment]
commented

File: app/views/users/index.html.erb (hunk 7 above)
Comment (TuanNhon) [reply app/views/users/index.html.erb]: Sorry, this make a easy sense for me. My team have another one make the interfaces

**Comment by TuanNhon** [comment]
ready

**Pull request merged**